queries/exchange.js

```diff
@@ -0,0 +1,425 @@
+                    next(results) {
+                        next(pairs.callback(results.data.pairs));
+                    },
+                    error,
+                    complete
+                });
+            }
+        };
+    },
+
+    async ethPrice({block = undefined, timestamp = undefined} = {}) {
+        block = block ? block : timestamp ? (await timestampToBlock(timestamp)) : undefined;
+        block = block ? `block: { number: ${block} }` : "";
+
+        const result = await request(graphAPIEndpoints.exchange,
+            gql`{
+                    bundle(id: 1, ${block}) {
+                        ${ethPrice.properties.toString()}
+                    }
+                }`
+        );
+
+        return ethPrice.callback(result.bundle);
+    },
+
+    observeEthPrice() {
+        const query = gql`
+            subscription {
+                bundle(id: 1) {
+                    ${ethPrice.properties.toString()}
+                }
+        }`;
+
+        const client = new SubscriptionClient(graphWSEndpoints.exchange, { reconnect: true, }, ws,);
+        const observable = client.request({ query });
+
+        return {
+            subscribe({next, error, complete}) {
+                return observable.subscribe({
+                    next(results) {
+                        next(ethPrice.callback(results.data.bundle));
+                    },
+                    error,
+                    complete
+                });
+            }
+        };
+    },
+
+    async factory({block = undefined, timestamp = undefined} = {}) {
+        block = block ? block : timestamp ? (await timestampToBlock(timestamp)) : undefined;
+        block = block ? `block: { number: ${block} }` : "";
+
+        const result = await request(graphAPIEndpoints.exchange,
+            gql`{
+                    factory(id: "0xc0aee478e3658e2610c5f7a4a2e1777ce9e4f2ac", ${block}) {
+                        ${factory.properties.toString()}
+                    }
+                }`
+        );
+
+        return factory.callback(result.factory);
+    },
+
+    observeFactory() {
+        const query = gql`
+            subscription {
+                factory(id: "0xc0aee478e3658e2610c5f7a4a2e1777ce9e4f2ac") {
+                    ${factory.properties.toString()}
+                }
+        }`;
+
+        const client = new SubscriptionClient(graphWSEndpoints.exchange, { reconnect: true, }, ws,);
+        const observable = client.request({ query });
+
+        return {
+            subscribe({next, error, complete}) {
+                return observable.subscribe({
+                    next(results) {
+                        next(factory.callback(results.data.factory));
+                    },
+                    error,
+                    complete
+                });
+            }
+        };
+    },
+
+    async dayData({minTimestamp = undefined, maxTimestamp = undefined, minBlock = undefined, maxBlock = undefined} = {}) {
```

Comment:
Let's add a `max` parameter to all of the queries that return a large number of results like this one.

```suggestion
    async dayData({minTimestamp = undefined, maxTimestamp = undefined, minBlock = undefined, maxBlock = undefined, max = undefined} = {}) {
```